[AutoDiff] Fix conformance checking in @differentiable attribute wrt parameters

rxwei opened the pull request:
1. When type-checking a `@differentiable` declaration attribute, reject `wrt:` parameters whose type's declaration conditionally conforms to `Differentiable` whilst the parameter type itself does not satisfy the conditional conformance requirements.
2. When the type checker infers parameters to differentiate with respect to, do not include ones whose type's declaration conditionally conforms to `Differentiable` whilst the parameter type itself does not satisfy the conditional conformance requirements.

Implementation note:
- `@differentiable` attribute conformance checking and differentiation parameter inference are now robust. `TypeChecker::conformsToProtocol` is used everywhere for conformance checking instead of `LookUpConformanceInModule`/`ModuleDecl::lookupConformance`.
- The double `candidate is missing attribute` diagnostic is fixed. The `finalize` function returns a `RequirementMatch` - `@differentiable` attribute requirement-witness conflicts should be diagnosed only if the result of `finalize` is viable.
- With the fix above, `@differentiable` attributes now print only once (when there is a true `@differentiable` attribute conflict) in the reproducer.

Fixes [TF-449](https://bugs.swift.org/browse/TF-449) and [TF-284](https://bugs.swift.org/browse/TF-284).

rxwei committed:
WIP

dan-zheng committed:
Fix contextual type mapping.

One failing test:

```
swift/test/AutoDiff/Inputs/differentiable_requirement_other_module.swift:4:4: error: can only differentiate with respect to parameters that conform to 'Differentiable', but 'Self' does not conform to 'Differentiable'
  @differentiable
   ^
```

dan-zheng committed:
Fix differentiation parameter inference in TypeCheckAttr.cpp.

Use same logic for inferring and checking differentiation parameters.

dan-zheng commented:
Ready for review. cc @rxwei (I can't add you as a reviewer on your own PR)

---

Friend API, which should be merged first: https://github.com/tensorflow/swift-apis/pull/194

dan-zheng commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2569,18 +2569,75 @@ void AttributeChecker::visitPropertyDelegateAttr(PropertyDelegateAttr *attr) {
+/// - If the function type's result is a function type, then also all
+///   parameters of the function result type that conform to `Differentiable`.
+static AutoDiffParameterIndices *
+inferParameters(AnyFunctionType *functionType, TypeChecker &TC,
```
It would be great to remove/deduplicate `AutoDiffParameterIndicesBuilder::inferParameters` if possible.
I removed one usage from `Builtins.cpp`, but there remains [a usage in `Attr.cpp`](https://github.com/apple/swift/blob/48f7718a3fb53701951c6ecd4c4595e3d3a7aeff/lib/AST/Attr.cpp#L351).
I'm not sure how to deduplicate, so I left `AutoDiffParameterIndicesBuilder::inferParameters` untouched.

cc @rxwei

rxwei replied:
I suspect `inferParameters`'s behavior is still incorrect for that usage. It's printing so even if it behaved incorrectly it might be okay. 

It would be quite a hassle to dedupe, but it'll be really great to make the attribute printing logic not depend on conformance checking. Instead, Sema should provide its own logic for pretty printing wrt parameters.

dan-zheng replied:
The only usage of pretty printing wrt parameters is in [`static void diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance, ValueDecl *req, const RequirementMatch &match)`](https://github.com/apple/swift/blob/48f7718a3fb53701951c6ecd4c4595e3d3a7aeff/lib/Sema/TypeCheckProtocol.cpp#L2251), in TypeCheckProtocol.cpp.

That function doesn't currently take a `TypeChecker &` instance, so we'd have to edit the function to take a `TypeChecker &` and move pretty printing logic out of `getDifferentiationParametersClauseString` in `Attr.cpp` and create another version taking a `TypeChecker &` in TypeCheckProtocol.cpp.

What are your thoughts? Shall we fix the attribute printing logic in this PR?

slavapestov replied:
It might not actually need a TypeChecker anymore. We've been making methods of TypeChecker static, but some places still use a TypeChecker instance while only calling static methods on it.

dan-zheng replied:
Removed `TypeChecker &` argument from functions in https://github.com/apple/swift/commit/38de6947fc62bc1c734d6f5f6f3359f3b5c93b13.

I'm still working on removing `AutoDiffParameterIndicesBuilder::inferParameters` and refactoring the pretty printing logic. Current progress is: there's a regression regarding [`differentiable_attr_type_checking.swift`](https://github.com/apple/swift/blob/ef437460ccfb561505dcaf4b39292522fa85bf32/test/AutoDiff/differentiable_attr_type_checking.swift#L669):

```swift
protocol DifferentiableInit : Differentiable {
  // expected-note @+2 {{protocol requires initializer 'init(x:y:)' with type '(x: Float, y: Float)'}}
  @differentiable
  init(x: Float, y: Float)

  // expected-note @+2 {{protocol requires initializer 'init(x:y:)' with type '(x: Float, y: Int)'}}
  @differentiable(wrt: x)
  init(x: Float, y: Int)
}
// expected-error @+1 {{does not conform to protocol 'DifferentiableInit'}}
struct DifferentiableInitStruct : DifferentiableInit {
  var x: Float
  var y: Float

  // FIXME(TF-284): Fix unexpected diagnostic.
  // expected-note @+2 {{candidate is missing attribute '@differentiable'}}
  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
  init(x: Float, y: Float) {
    self.x = x
    self.y = y
  }

  // FIXME(TF-284): Fix unexpected diagnostic.
  // expected-note @+2 {{candidate is missing attribute '@differentiable'}}
  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
  init(x: Float, y: Int) {
    self.x = x
    self.y = Float(y)
  }
}
```

```
/Users/danielzheng/swift-dev/swift/test/AutoDiff/differentiable_attr_type_checking.swift:670:26: error: incorrect message found
  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         candidate is missing attribute '@differentiable(wrt: x)'
/Users/danielzheng/swift-dev/swift/test/AutoDiff/differentiable_attr_type_checking.swift:678:26: error: incorrect message found
  // expected-note @+1 {{candidate is missing attribute '@differentiable'}}
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         candidate is missing attribute '@differentiable(wrt: (x, y))'
/Users/danielzheng/swift-dev/swift/test/AutoDiff/differentiable_attr_type_checking.swift:694:26: error: incorrect message found
  // expected-note @+1 {{candidate is missing attribute '@differentiable(wrt: x)'}}
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                         candidate is missing attribute '@differentiable'
/Users/danielzheng/swift-dev/swift/test/AutoDiff/differentiable_attr_type_checking.swift:710:6: error: expected note not produced
  // expected-note @+2 {{candidate is missing attribute '@differentiable(wrt: x)}}
~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

There are two unexpected things here:
* [TF-284](https://bugs.swift.org/projects/TF/issues/TF-284): There are double `candidatte is missing attribute` errors on both witness initializers in `struct DifferentiableInitStruct`. This seems to occur because [`WitnessChecker::lookupValueWitnesses(ValueDecl *req, bool *ignoringNames)`](https://github.com/apple/swift/blob/ef437460ccfb561505dcaf4b39292522fa85bf32/lib/Sema/TypeCheckProtocol.cpp#L1068) returns two witnesses for each `init(x:y:)` requirement in `protocol DifferentiableInit`, which is unexpected because only one of the witnesses has a matching type. In other words, it seems weird `DifferentiableInitStruct.init(x: Float, y: Int)` is returned as a witness for `DifferentiableInit.init(x: Float, y: Float)`. cc @slavapestov: do you happen to know why this occurs?
* The regression is that attributes now print differently (e.g. once as `@differentiable` and once as `@differentiable(wrt: (x, y))`).

dan-zheng replied:
Update: I found fixes regarding both issues! Will edit this comment with more information.

rxwei commented on lib/AST/Builtins.cpp:
```diff
@@ -1024,9 +1024,10 @@ static ValueDecl *getAutoDiffApplyAssociatedFunction(
-  auto autodiffBuilder = AutoDiffParameterIndicesBuilder::inferParameters(
-      origFnTy, Context.getStdlibModule());
-  auto *paramIndices = autodiffBuilder.build(Context);
+  AutoDiffParameterIndicesBuilder autoDiffBuilder(origFnTy);
```
This isn't the best way to do this, because `AutoDiffIndexSubset` provides a `getDefault()` factory method that is more efficient and easier to use.

```cpp
auto *paramIndices =
    AutoDiffIndexSubset::getDefault(Context, origFnTy->getNumParams(), /*includeAll*/ true);
```

dan-zheng replied:
`paramIndices` here has type `AutoDiffParameterIndices *`, which is necessary as an argument to `AnyFunctionType::getAutoDiffAssociatedFunctionType`. I don't believe `AutoDiffIndexSubst *` is suitable?

rxwei replied:
Oh shoot, I thought my #25140 had landed.

In this case, use `AutoDiffParameterIndices::get(SmallBitVector(origFnTy->getNumParams(), true), Context)`.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/25324/commits/3313a81d9762ab44447ca3f170874479d4de5a5e.

dan-zheng committed:
Merge remote-tracking branch 'apple/tensorflow' into differentiable-wrt-conformance

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2569,18 +2569,75 @@ void AttributeChecker::visitPropertyDelegateAttr(PropertyDelegateAttr *attr) {
+/// - If the function type's result is a function type, then also all
+///   parameters of the function result type that conform to `Differentiable`.
+static AutoDiffParameterIndices *
+inferParameters(AnyFunctionType *functionType, TypeChecker &TC,
```
The name here is a bit too general for `TypeCheckAttr.cpp`. How about `inferDifferentiableParameters`?

dan-zheng committed:
Re-add deserialization hack.

Reverts change in https://github.com/apple/swift/pull/25293.

rxwei commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2569,18 +2569,75 @@ void AttributeChecker::visitPropertyDelegateAttr(PropertyDelegateAttr *attr) {
+/// - If the function type's result is a function type, then also all
+///   parameters of the function result type that conform to `Differentiable`.
+static AutoDiffParameterIndices *
+inferParameters(AnyFunctionType *functionType, TypeChecker &TC,
```
How about just taking `(TypeChecker &, AbstractFunctionDecl *, GenericEnvironment *)`? You can get the decl context and the function type from the function decl.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/25324/commits/3313a81d9762ab44447ca3f170874479d4de5a5e.

rxwei reviewed (commented):
Thanks for taking over this fix!

dan-zheng committed:
Address review comments.

rxwei commented on lib/AST/Builtins.cpp:
```diff
@@ -1027,7 +1027,8 @@ static ValueDecl *getAutoDiffApplyAssociatedFunction(
   AutoDiffParameterIndicesBuilder autoDiffBuilder(origFnTy);
```
And 1027-1029 should be removed since they are no longer used.

dan-zheng replied:
Oops, done in https://github.com/apple/swift/pull/25324/commits/866c6abe948bb44fb27c0d66957d2184fab16249.

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2569,18 +2569,78 @@ void AttributeChecker::visitPropertyDelegateAttr(PropertyDelegateAttr *attr) {
+      return false;
+    auto paramType = allParamTypes[i];
+    if (!paramType->hasTypeParameter())
+      paramType = paramType->mapTypeOutOfContext();
```
These two lines are a no-op. mapTypeOutOfContext() substitutes type parameters, but you're only calling it if the type has no type parameters. You should just be able to delete this.

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/25324/commits/38de6947fc62bc1c734d6f5f6f3359f3b5c93b13. I'll try to avoid this no-op pattern in the future.

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2569,18 +2569,78 @@ void AttributeChecker::visitPropertyDelegateAttr(PropertyDelegateAttr *attr) {
+    else
+      paramType = AFD->mapTypeIntoContext(paramType);
+    // Return false for class/existential types.
+    if ((!paramType->hasTypeParameter() &&
```
!hasTypeParameter is a post-condition of mapTypeIntoContext(), so part of this boolean check is also always false.

dan-zheng replied:
Done in https://github.com/apple/swift/commit/38de6947fc62bc1c734d6f5f6f3359f3b5c93b13.

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3053,15 +3103,16 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
-    auto originalResultInterfaceType = !originalResultTy->hasTypeParameter()
-        ? originalResultTy->mapTypeOutOfContext()
-        : originalResultTy;
+  originalResultTy = originalResultTy->hasTypeParameter()
```
For the same reasons, this expression will never change originalResultTy.

dan-zheng replied:
Done in https://github.com/apple/swift/commit/38de6947fc62bc1c734d6f5f6f3359f3b5c93b13.

dan-zheng committed:
Address review comments.

- Remove unnecessary type remapping.
- Remove unnecessary `TypeChecker &` argument from functions.

dan-zheng committed:
Refactor `@differentiable` attribute pretty-printing logic.

- Change `DifferentiableAttr::print` to take `AutoDiffParameterIndices *`
  inferred parameters for pretty printing.
- Update `@differentiable` attribute pretty-printing logic in
  TypeCheckProtocol.cpp.
- Remove `AutoDiffParameterIndicesBuilder::inferParameters`.

dan-zheng committed:
Remove unused function.

dan-zheng commented:
Ready for re-review. Final fix documented in [this comment](https://github.com/apple/swift/pull/25324#discussion_r292562968).

`@differentiable` attribute conformance checking and differentiation parameter inference are now robust. `TypeChecker::conformsToProtocol` is used everywhere for conformance checking instead of `LookUpConformanceInModule`/`module->lookupConformance`.

rxwei commented on include/swift/AST/Attr.h:
```diff
@@ -1595,7 +1595,8 @@ class DifferentiableAttr final
   // Print the attribute to the given stream.
   void print(llvm::raw_ostream &OS, const Decl *D,
-             ModuleDecl *prettyPrintInModule = nullptr) const;
+             AutoDiffParameterIndices *inferredParametersForPrettyPrinting =
```
Is it the case that the only thing `print()` will do is to check whether `inferredParametersForPrettyPrinting->parameters.count()` is equal to the actual count so that it will not print it?  In that case, I find passing an `AutoDiffParameterIndices *` argument a bit unnecessary.  It should suffice to just take a Boolean flag (e.g. `omitWRT`) that indicates whether the `wrt:` clause in the attribute should be printed.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/25324/commits/7759a3db53cb7ad24776de7a7f0b241c8d44781d. Commit description:

```
Use `omitDifferentiationParametersClause` boolean flag.

Caller in TypeCheckProtocol.cpp computes the flag.
Create helper function if there become more callers in Sema.
```

I tried creating a `std::string getDifferentiableAttributeString(const DifferentiableAttr *attr, AbstractFunctionDecl *original)` helper in Sema, but that didn't work because it requires two `AbstractFunctionDecl *` arguments: one for inferring parameters (witness) and one for printing the attribute (requirement). I decided not to create the helper for now.

dan-zheng committed:
Use `omitDifferentiationParametersClause` boolean flag.

Caller in TypeCheckProtocol.cpp computes the flag.
Create helper function if there become more callers in Sema.

rxwei commented on include/swift/AST/Attr.h:
```diff
@@ -1595,7 +1595,7 @@ class DifferentiableAttr final
   // Print the attribute to the given stream.
   void print(llvm::raw_ostream &OS, const Decl *D,
-             ModuleDecl *prettyPrintInModule = nullptr) const;
+             bool omitDifferentiationParametersClause = false) const;
```
`omitWRTClause` is shorter and more directly relatable.

dan-zheng replied:
How about `omitWrtClause`?

dan-zheng replied:
Renamed to `omitWrtClause` in https://github.com/apple/swift/pull/25324/commits/12af26bb99bf20544df43cbc6a78e026c697ca05.

rxwei reviewed (commented):
LGTM. Thanks a lot for pushing this PR to a final fix.

dan-zheng committed:
Rename to `omitWrtClause`.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into differentiable-wrt-conformance

dan-zheng committed:
Minor fixes.

- Remove usage of `ConformanceCheckFlags::Used`.
- Update tensorflow-swift-apis hash.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow
